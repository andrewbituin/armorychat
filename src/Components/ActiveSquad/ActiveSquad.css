.active-squad-container {
  width: 500px;
  padding: 0 20px;
}

.active-squad__details {
  display: flex;
}

.active-squad__details-text-container {
  padding-top: 10px;
}

.active-squad__details-text-container p, .active-squad__details-text-container span {
  margin: 10px 0;
}

.active-squad__details-text-container span {
  margin-right: 10px;
}

.active-squad__details-text-container p {
  width: 360px;
  text-overflow: ellipsis;
  overflow: hidden;
  word-wrap: break-word;
  max-height: 60px;
}

.active-squad__section-header {
  font-size: 14px;
  display: block;
  font-weight: normal;
  border-bottom: 1px solid #5b6370;
  margin: 10px 0;
}

.active-squad__owner {
  display: flex;
  align-items: center;
  width: 200px;
  padding-bottom: 5px;
  padding-left: 10px;
}

.active-squad__owner div {
  margin-right: 10px;
}

.active-squad__fake-checkbox, .active-squad__checkbox {
  box-sizing: content-box;
  position: relative;
  width: 40px;
  height: 40px;
  border-radius: 5px;
  overflow: hidden;
  box-shadow: 0 1px 8px 0 #00000080;
  border: 1px solid #909193;
}

.active-squad__checkbox-container {
  width: 60px;
  height: 50px;
  display: inline-flex;
  justify-content: center;
  align-items: center;
}

.active-squad__checkbox {
  background-color: #000;
  display: inline-block;
}

.active-squad__checkbox:hover {
  cursor: pointer;
}

.active-squad__active-checkbox {
  border: 4px solid #6fa42c;
}

.active-squad__spots {
  display: flex;
  flex-direction: column;
  width: 100%;
}

.active-squad__spots-list {
  list-style: none;
  display: flex;
  justify-content: space-around;
  padding: 0;
}

.active-squad__spots-full, .active-squad__spots-first-half, .active-squad__spots-second-half {
  position: absolute;
  left: -10px;
  top: -10px;
  width: 60px;
  height: 60px;
}

.active-squad__spots-first-half {
  clip-path: inset(10px 30px 10px 10px);
}

.active-squad__spots-second-half {
  clip-path: inset(10px 10px 10px 30px);
}

.active-squad__submit-button {
  width: 45%;
  height: 30px;
}

.active-squad__cancel-button {
  width: 30%;
  height: 30px;
}

@media screen and (max-width: 900px) {
  .active-squad-container {
    width: 100%;
    height: 100%;
  }
  
  .active-squad__details-text-container p {
    width: 100%;
    max-height: unset;
  }
}