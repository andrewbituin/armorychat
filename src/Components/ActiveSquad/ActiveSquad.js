import React, { useState, useContext } from 'react';
import config from '../../config';
import './ActiveSquad.css';
import TokenService from '../../services/token-service';
import GameContext from '../../Contexts/gameContext';

export default function ActiveSquad(props) {
  const gameContext = useContext(GameContext);
  const [spotToJoin, setSpotToJoin] = useState(undefined);
 
  function generateRoleImages(roles, filled) {
    let temp;
    if (roles.length === 1) {
      if (roles[0].icon_url) {
        temp = (
          <img className="active-squad__spots-full" src={`${config.IMAGES_ENDPOINT}/${gameContext.id}/${roles[0].icon_url}`} alt=""/>
        );
      } else {
        temp = (<i className="squad__spots-empty fas fa-users"/>);
      }
    } else if (roles.length > 1) {
      temp = (
        <>
          <img className="active-squad__spots-first-half" src={`${config.IMAGES_ENDPOINT}/${gameContext.id}/${roles[0].icon_url}`} alt=""/>
          <img className="active-squad__spots-second-half" src={`${config.IMAGES_ENDPOINT}/${gameContext.id}/${roles[1].icon_url}`} alt=""/>
        </>
      );
    }
    if (filled) {
      return (
        <>
          {temp}
          <i className="squad__spots-filled fas fa-check"/>
        </>
      );
    } else {
      return temp;
    } 
  }

  function handleSpotOnClick(index) {
    if (spotToJoin === index) {
      setSpotToJoin(undefined);
    } else {
      setSpotToJoin(index);
    }
  }

  function generateSpots(party) {
    return party.spots
      .map((spot, index) => {
        if (spot.filled) {
          return <li key={index} className="active-squad__checkbox-container">
            <div className="active-squad__fake-checkbox">
              {generateRoleImages(spot.roles, spot.filled)}
            </div>
          </li>
        } else {
          return <li 
            className="active-squad__checkbox-container"
            key={index}
          >
            <div
              className={index === spotToJoin ? "active-squad__checkbox active-squad__active-checkbox" : "active-squad__checkbox"} 
              tabIndex="0" 
              aria-label="checkbox"
              onClick={() => handleSpotOnClick(index)}
            >
              {generateRoleImages(spot.roles, spot.filled)}
            </div>
          </li>;
        }
      });
  }

  function generateRequirements() {
    return props.party.reqs.map((req, i) => {
      if (req.name) {
        return <span className="small-detail" key={i}>{req.name}</span>
      }
    });
  }

  function handleJoinSpot(e) {
    e.preventDefault();
    fetch(`${config.API_ENDPOINT}/spot/${props.party.spots[spotToJoin].id}`, {
      method: 'PATCH',
      headers: {
        authorization: `Bearer ${TokenService.getAuthToken()}`,
        'content-type': 'application/json'
      },
      body: JSON.stringify({ game_id: gameContext.id, party_id: props.party.id })
    })
      .then(res => {
        if (!res.ok) {
          return res.json().then(e => Promise.reject(e));
        } else {
          props.history.push(`/party/${props.party.id}`);
        }
      })
      .catch(err => {
        gameContext.setError(err);
      });
  }

  return (
    <div className="active-squad-container">
      <h2 className="modal-header">{props.party.title}</h2>
      <span className="active-squad__section-header">Details</span>
      <div className="active-squad__details">
        <div className="squad__gamemode-image-container">
          {(props.party.gamemode) 
            ? <img className="squad__gamemode-image" src={`${config.IMAGES_ENDPOINT}/${gameContext.id}/${props.party.gamemode.icon_url}`} alt=""/>
            : <i className="squad__gamemode-image fas fa-globe squad__gamemode-image-icon"></i>
          }
        </div>
        <div className="active-squad__details-text-container">
          {props.party.gamemode && <span className="small-detail">{props.party.gamemode.name}</span>}
          {generateRequirements()}
          <p>{props.party.description || 'No description...'}</p>
        </div>
      </div>
      <span className="active-squad__section-header">Owner</span>
      <div className="active-squad__owner">
        <div className="squad__spots-image-container">
          <img className="squad__avatar-image" src={`${config.IMAGES_ENDPOINT}/user-icons/${props.party.owner_id.avatar_url}`} alt=""/>
        </div>
        <p>{props.party.owner_id.username}</p>
      </div>
      <span className="active-squad__section-header">Select a spot...</span>
      <form className="active-squad__spots" onSubmit={handleJoinSpot}>
        <ul className="active-squad__spots-list">
          {generateSpots(props.party)}
        </ul>
        <div className="create-squad__button-container">
          <button className="green-button active-squad__submit-button" type="submit" disabled={!spotToJoin}>Join Squad</button>
          <button className="grey-button active-squad__cancel-button" type="button" aria-label="cancel" onClick={props.hide}>Cancel</button>
        </div>
      </form>
    </div>
  )
}