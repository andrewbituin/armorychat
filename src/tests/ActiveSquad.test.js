import React from "react";
import ActiveSquad from "../Components/ActiveSquad/ActiveSquad";
import { MemoryRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import { mount } from "enzyme";

const props = {
  index: 1,
  id: "fb1d3c63-6a72-4013-be82-5b523c1dd1cd",
  game_id: "aa0e8ce9-1a71-42e7-804d-6838556fa6ed",
  title: "Admin",
  owner_id: 1,
  description: "This is a description of this party.",
  require_app: true,
  gamemode: {
    name: "Quick Play",
    icon_url: "Gamemode-Quickplay.png"
  },
  reqs: [
    {
      name: "Silver"
    },
    {
      name: "Bronze"
    }
  ],
  spots: [
    {
      id: "64ed5ba8-78db-44c6-ae60-46e6a2a07ff9",
      filled: {
        username: "admin",
        avatar_url: "https://i.ebayimg.com/images/g/PfAAAOSwA3dYIPRN/s-l300.jpg"
      },
      roles: [
        {
          name: "Genji",
          icon_url: "Icon-Genji.png"
        },
        {
          name: "Support",
          icon_url: "Icon-Support.png"
        }
      ]
    }
  ]
};

describe("Game component", () => {
  // remove unique key warning message
  let savedError;
  beforeEach(() => {
    savedError = console.error;
    console.error = jest.fn();
  })

  afterEach(() => {
    console.error = savedError;
  })

  it("renders without crashing", () => {
    mount(<ActiveSquad party={props} />);
  });

  it("renders the UI as expected", () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <ActiveSquad party={props} />
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
