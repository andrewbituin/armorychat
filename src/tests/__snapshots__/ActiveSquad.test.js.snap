// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game component renders the UI as expected 1`] = `
<div
  className="active-squad-container"
>
  <h2
    className="modal-header"
  >
    Admin
  </h2>
  <span
    className="active-squad__section-header"
  >
    Details
  </span>
  <div
    className="active-squad__details"
  >
    <div
      className="squad__gamemode-image-container"
    >
      <img
        alt=""
        className="squad__gamemode-image"
        src="https://limitless-brushlands-45977.herokuapp.com/images/[object Object]/Gamemode-Quickplay.png"
      />
    </div>
    <div
      className="active-squad__details-text-container"
    >
      <span
        className="small-detail"
      >
        Quick Play
      </span>
      <span
        className="small-detail"
      >
        Silver
      </span>
      <span
        className="small-detail"
      >
        Bronze
      </span>
      <p>
        This is a description of this party.
      </p>
    </div>
  </div>
  <span
    className="active-squad__section-header"
  >
    Owner
  </span>
  <div
    className="active-squad__owner"
  >
    <div
      className="squad__spots-image-container"
    >
      <img
        alt=""
        className="squad__avatar-image"
        src="https://limitless-brushlands-45977.herokuapp.com/images/user-icons/undefined"
      />
    </div>
    <p />
  </div>
  <span
    className="active-squad__section-header"
  >
    Select a spot...
  </span>
  <form
    className="active-squad__spots"
    onSubmit={[Function]}
  >
    <ul
      className="active-squad__spots-list"
    >
      <li
        className="active-squad__checkbox-container"
      >
        <div
          className="active-squad__fake-checkbox"
        >
          <img
            alt=""
            className="active-squad__spots-first-half"
            src="https://limitless-brushlands-45977.herokuapp.com/images/[object Object]/Icon-Genji.png"
          />
          <img
            alt=""
            className="active-squad__spots-second-half"
            src="https://limitless-brushlands-45977.herokuapp.com/images/[object Object]/Icon-Support.png"
          />
          <i
            className="squad__spots-filled fas fa-check"
          />
        </div>
      </li>
    </ul>
    <div
      className="create-squad__button-container"
    >
      <button
        className="green-button active-squad__submit-button"
        disabled={true}
        type="submit"
      >
        Join Squad
      </button>
      <button
        aria-label="cancel"
        className="grey-button active-squad__cancel-button"
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
</div>
`;
